### PR TITLE
v14: Make state.open nullable again

### LINKED ISSUE #86: state.open cannot be undefined when supporting both v13 and v0.14
In v0.13 the `state.open` field was required but nullable. In v14 that was changed, the field is now optional but not nullable.

Unfortunately that means that it's not possible to have a v0.13+v14-compliant endpoint that sets the opening state to "undefined" :confused: 

Do we need to take action, or do we just live with that fact for now?

### PR DESCRIPTION
**Pull request by rnestler**
This is for better backwards compatibility with v0.13.

Fixes #86

### CODE REVIEW

File: 14.json [hunk 1]
```diff
@@ -83,8 +83,11 @@
       "type": "object",
       "properties": {
         "open": {
-          "description": "A flag which indicates whether the space is currently open or closed. The state 'undefined' can be achieved by omitting this field. A missing 'open' property carries the semantics of a temporary unavailability of the state, whereas the absence of the 'state' property itself means the state is generally not implemented for this space.",
-          "type": "boolean"
+          "description": "A flag which indicates whether the space is currently open or closed. The state 'undefined' can be achieved by omitting this field. A missing 'open' property carries the semantics of a temporary unavailability of the state, whereas the absence of the 'state' property itself means the state is generally not implemented for this space. This field is also allowed to explicitly have the value null for backwards compatibility with older schema versions.",
```
Comment (dbrgn) [inline 14.json:86]: Should we add a deprecation note? Endpoints targeting only v14 or higher should not set the field to null.

Comment (rnestler) [reply 14.json:86]: How would the deprecation note look like?

Comment (dbrgn) [reply 14.json:86]: Simply stating in the description that the field is deprecated 🙂 (I don't think there's a formal mechanism.)

Comment (rnestler) [reply 14.json:86]: Well the field is not deprecated, just setting it to `null` is, but I guess this is what you meant.

Comment (rnestler) [reply 14.json]: I mentioned that it is deprecated.

**Review by dbrgn — APPROVED** [review]
LGTM!

**Review by gidsi — COMMENTED** [review]
Only problem i have with that situation is that we're changing an agreed "standard", doesn't feel nice/right. 

How can we make sure that this is a one time thing and not a broken window situation?

**Comment by rnestler** [comment]
> Only problem i have with that situation is that we're changing an agreed "standard", doesn't feel nice/right.

IMO fixing bugs in the agreed "standard" should be allowed. The question is more what exactly counts as bug fixing. I guess fixing typos in the documentation of the schema we'd all agree is OK. Renaming a key probably not (even if it has a typo in the key name).

I'd argue that the *intended* change we wanted to do in v14 is to remove `state.open` as a required field so that there is no *need* to set it to `null`. But we probably didn't want to make life for users hard by having contradicting requirements in v0.13 and v14, since we explicitly want to allow to support multiple versions of the API at the same time.

So I consider this a bug fix to change it the way we intended it to be.

**Comment by rnestler** [comment]
> How can we make sure that this is a one time thing and not a broken window situation?

I agree we shouldn't make changes to old versions of the standard a regular thing. I would propose the following limitations:
 * Only changing of the latest published standard is allowed. (So no changing of v0.13 at this point)
 * Only changes which fix bugs which make the life of users hard for no apparent reason
 * Fixes must be fully backward compatible (All valid documents before must be valid after the change)

**Comment by gidsi** [comment]
> > How can we make sure that this is a one time thing and not a broken window situation?
> 
> I agree we shouldn't make changes to old versions of the standard a regular thing. I would propose the following limitations:
> 
>     * Only changing of the latest published standard is allowed. (So no changing of v0.13 at this point)
> 
>     * Only changes which fix bugs which make the life of users hard for no apparent reason
> 
>     * Fixes must be fully backward compatible (All valid documents before must be valid after the change)

But they are not backward compatible at this point, thats my issue.

Generated types and things like this are going to break too (which im fine with, but we shouldnt do it again).

But i dont have any idea how we wont get in the situation again beside "work better"

**Review by gidsi — APPROVED** [review]

**Comment by rnestler** [comment]
> But i dont have any idea how we wont get in the situation again beside "work better"

I guess we should actually update a reference implementation even for not yet published schema versions. That way we would notice practical problems with the schema immediately.

**Pull request merged**